// fbmessenger/dist/templates/ReceiptTemplate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
class ReceiptTemplate {
  constructor(_ref) {
    var recipient_name = _ref.recipient_name,
        order_number = _ref.order_number,
        currency = _ref.currency,
        payment_method = _ref.payment_method,
        elements = _ref.elements,
        summary = _ref.summary,
        _ref$timestamp = _ref.timestamp,
        timestamp = _ref$timestamp === undefined ? '' : _ref$timestamp,
        _ref$order_url = _ref.order_url,
        order_url = _ref$order_url === undefined ? '' : _ref$order_url,
        _ref$address = _ref.address,
        address = _ref$address === undefined ? '' : _ref$address,
        _ref$adjustments = _ref.adjustments,
        adjustments = _ref$adjustments === undefined ? [] : _ref$adjustments;

    this.recipient_name = recipient_name;
    this.order_number = order_number;
    this.currency = currency;
    this.payment_method = payment_method;
    this.elements = elements;
    this.summary = summary;
    this.timestamp = timestamp;
    this.order_url = order_url;
    this.address = address;
    this.adjustments = adjustments;

    var res = {
      attachment: {
        type: 'template',
        payload: {
          template_type: 'receipt',
          recipient_name: this.recipient_name,
          order_number: this.order_number,
          currency: this.currency,
          payment_method: this.payment_method,
          order_url: this.order_url,
          timestamp: this.timestamp,
          elements: this.elements,
          address: this.address,
          summary: this.summary
        }
      }
    };

    if (this.adjustments.length) {
      res.attachment.payload.adjustments = this.adjustments;
    }

    return res;
  }
}

exports.default = ReceiptTemplate;
