// fbmessenger/dist/elements/Button.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _constants = require('../constants');

class Button {
  constructor(_ref) {
    var type = _ref.type,
        title = _ref.title,
        _ref$url = _ref.url,
        url = _ref$url === undefined ? '' : _ref$url,
        _ref$payload = _ref.payload,
        payload = _ref$payload === undefined ? '' : _ref$payload,
        _ref$webview_height_r = _ref.webview_height_ratio,
        webview_height_ratio = _ref$webview_height_r === undefined ? '' : _ref$webview_height_r,
        _ref$messenger_extens = _ref.messenger_extensions,
        messenger_extensions = _ref$messenger_extens === undefined ? '' : _ref$messenger_extens,
        _ref$fallback_url = _ref.fallback_url,
        fallback_url = _ref$fallback_url === undefined ? '' : _ref$fallback_url;

    if (_constants.BUTTON_TYPES.indexOf(type) === -1) {
      throw new Error('Invalid button type provided.');
    }

    if (webview_height_ratio && _constants.WEBVIEW_HEIGHT_RATIOS.indexOf(webview_height_ratio) === -1) {
      throw new Error('Invalid webview_height_ratio provided.');
    }

    if (title.length > 20) {
      throw new Error('Title cannot be longer 20 characters.');
    }

    this.type = type;
    this.title = title;
    this.url = url;
    this.payload = payload;
    this.webview_height_ratio = webview_height_ratio;
    this.messenger_extensions = messenger_extensions;
    this.fallback_url = fallback_url;

    var button = {
      type: this.type,
      title: this.title
    };

    if (this.url) {
      button.url = this.url;
    }

    if (this.payload) {
      button.payload = this.payload;
    }

    if (this.webview_height_ratio) {
      button.webview_height_ratio = this.webview_height_ratio;
    }

    if (this.messenger_extensions) {
      button.messenger_extensions = 'true';
    }

    if (this.fallback_url) {
      button.fallback_url = this.fallback_url;
    }

    return button;
  }
}

exports.default = Button;
