// fbmessenger/dist/elements/Element.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
class Element {
  constructor(_ref) {
    var title = _ref.title,
        _ref$item_url = _ref.item_url,
        item_url = _ref$item_url === undefined ? '' : _ref$item_url,
        _ref$image_url = _ref.image_url,
        image_url = _ref$image_url === undefined ? '' : _ref$image_url,
        _ref$subtitle = _ref.subtitle,
        subtitle = _ref$subtitle === undefined ? '' : _ref$subtitle,
        _ref$quantity = _ref.quantity,
        quantity = _ref$quantity === undefined ? '' : _ref$quantity,
        _ref$price = _ref.price,
        price = _ref$price === undefined ? '' : _ref$price,
        _ref$currency = _ref.currency,
        currency = _ref$currency === undefined ? '' : _ref$currency,
        _ref$buttons = _ref.buttons,
        buttons = _ref$buttons === undefined ? [] : _ref$buttons,
        _ref$default_action = _ref.default_action,
        default_action = _ref$default_action === undefined ? '' : _ref$default_action;

    this.title = title;
    this.item_url = item_url;
    this.image_url = image_url;
    this.subtitle = subtitle;
    this.quantity = quantity;
    this.price = price;
    this.currency = currency;
    this.buttons = buttons;
    this.default_action = default_action;

    var res = {
      title: this.title,
      image_url: this.image_url,
      subtitle: this.subtitle
    };

    if (this.item_url) {
      res.item_url = this.item_url;
    }

    if (this.quantity) {
      res.quantity = this.quantity;
    }

    if (this.currency) {
      res.currency = this.currency;
    }

    if (this.price) {
      res.price = this.price;
    }

    if (this.buttons.length) {
      res.buttons = this.buttons;
    }

    if (this.default_action) {
      res.default_action = this.default_action;
    }

    return res;
  }
}

exports.default = Element;
