// react-facebook/lib/Feed.js
'use strict';

exports.__esModule = true;
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _qs = require('qs');

var _qs2 = _interopRequireDefault(_qs);

var _FacebookProvider = require('./FacebookProvider');

var _FacebookProvider2 = _interopRequireDefault(_FacebookProvider);

var _getCurrentHref = require('./utils/getCurrentHref');

var _getCurrentHref2 = _interopRequireDefault(_getCurrentHref);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

var Feed = (_temp2 = _class = function (_Component) {
  _inherits(Feed, _Component);

  function Feed() {
    var _temp, _this, _ret;

    _classCallCheck(this, Feed);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _Component.call.apply(_Component, [this].concat(args))), _this), _this.onClick = function (evn) {
      evn.preventDefault();
      evn.stopPropagation();

      var href = _this.getSharerHref();
      var _this$props = _this.props,
          width = _this$props.width,
          height = _this$props.height;


      var halfWidth = Math.floor(width / 2);
      var halfHeight = Math.floor(height / 2);

      var left = Math.floor(window.innerWidth / 2 - halfWidth);
      var top = Math.floor(window.innerHeight / 2 - halfHeight);

      var params = 'status=0, width=' + width + ', height=' + height + ', top=' + top + ', left=' + left + ', toolbar=0, location=0, menubar=0, directories=0, scrollbars=0';

      window.open(href, 'sharer', params);

      var children = _this.props.children;

      if (children && children.props && children.props.onClick) {
        children.props.onClick(evn);
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  Feed.prototype.getSharerHref = function getSharerHref() {
    var facebook = this.context.facebook;
    var _props = this.props,
        _props$appId = _props.appId,
        appId = _props$appId === undefined ? facebook.props.appID : _props$appId,
        redirectURI = _props.redirectURI,
        display = _props.display,
        from = _props.from,
        to = _props.to,
        _props$link = _props.link,
        link = _props$link === undefined ? (0, _getCurrentHref2.default)() : _props$link,
        picture = _props.picture,
        source = _props.source,
        name = _props.name,
        caption = _props.caption,
        description = _props.description,
        ref = _props.ref;


    var query = _qs2.default.stringify({
      app_id: appId,
      redirect_uri: redirectURI,
      display: display,
      from: from,
      to: to,
      link: link,
      picture: picture,
      source: source,
      name: name,
      caption: caption,
      description: description,
      ref: ref
    });

    return '//www.facebook.com/dialog/feed?' + query;
  };

  Feed.prototype.render = function render() {
    var children = this.props.children;


    return _react2.default.cloneElement(children, { onClick: this.onClick });
  };

  return Feed;
}(_react.Component), _class.contextTypes = _extends({}, _FacebookProvider2.default.childContextTypes), _class.propTypes = {
  children: _react.PropTypes.node,
  width: _react.PropTypes.number.isRequired,
  height: _react.PropTypes.number.isRequired,
  appId: _react.PropTypes.string,
  redirectURI: _react.PropTypes.string,
  display: _react.PropTypes.string.isRequired,
  from: _react.PropTypes.string,
  to: _react.PropTypes.string,
  link: _react.PropTypes.string,
  picture: _react.PropTypes.string,
  source: _react.PropTypes.string,
  name: _react.PropTypes.string,
  caption: _react.PropTypes.string,
  description: _react.PropTypes.string,
  ref: _react.PropTypes.string
}, _class.defaultProps = {
  display: 'popup',
  width: 626,
  height: 436,
  buttonClassName: 'btn btn-lg',
  iconClassName: 'fa fa-facebook pull-left',
  icon: true
}, _temp2);
exports.default = Feed;
